consensus/src/chained_bft/block_storage/pending_votes.rs

```diff
@@ -0,0 +1,175 @@
+pub struct PendingVotes {
+    /// `li_digest_to_votes` might keep multiple LedgerInfos per proposed block in order
+    /// to tolerate non-determinism in execution: given a proposal, a QuorumCertificate is going
+    /// to be collected only for all the votes that carry identical LedgerInfo.
+    /// LedgerInfo digest covers the potential commit ids, as well as the vote information
+    /// (including the 3-chain of a voted proposal).
+    /// Thus, the structure of `li_digest_to_votes` is as follows:
+    /// HashMap<ledger_info_digest, LedgerInfoWithSignatures>
+    li_digest_to_votes: HashMap<HashValue, LedgerInfoWithSignatures>,
+    /// Tracks all the signatures of the votes for the given round. In case we succeed to
+    /// aggregate 2f+1 signatures for the same round a TimeoutCertificate is formed.
+    /// Note that QuorumCert has higher priority than TimeoutCertificate (in case 2f+1 votes are
+    /// gathered for the same ledger info we are going to generate QuorumCert and not the
+    /// TimeoutCertificate).
+    round_to_tc: HashMap<Round, TimeoutCertificate>,
+    /// Map of Author to last vote info. Any pending vote from Author is cleaned up
+    /// whenever new vote is added by same Author
+    author_to_last_voted_info: HashMap<Author, LastVoteInfo>,
+}
+
+impl PendingVotes {
+    pub fn new() -> Self {
+        PendingVotes {
+            li_digest_to_votes: HashMap::new(),
+            round_to_tc: HashMap::new(),
+            author_to_last_voted_info: HashMap::new(),
+        }
+    }
+
+    pub fn insert_vote(
+        &mut self,
+        vote_msg: &VoteMsg,
+        min_votes_for_qc: usize,
+    ) -> VoteReceptionResult {
+        if let Err(e) = self.check_vote_valid(vote_msg) {
+            return e;
+        }
+        let vote_aggr_res = self.aggregate_qc(vote_msg, min_votes_for_qc);
+        if let VoteReceptionResult::NewQuorumCertificate(_) = vote_aggr_res {
+            return vote_aggr_res;
+        }
+        match self.aggregate_tc(vote_msg, min_votes_for_qc) {
+            Some(VoteReceptionResult::NewTimeoutCertificate(tc)) => {
+                VoteReceptionResult::NewTimeoutCertificate(tc)
+            }
+            _ => vote_aggr_res,
+        }
+    }
+
+    /// Check whether the newly inserted vote completes a QC
+    fn aggregate_qc(&mut self, vote_msg: &VoteMsg, min_votes_for_qc: usize) -> VoteReceptionResult {
+        // Note that the digest covers the ledger info information, which is also indirectly
+        // covering vote data hash (in its `consensus_data_hash` field).
+        let li_digest = vote_msg.ledger_info().hash();
+        let li_with_sig = self.li_digest_to_votes.entry(li_digest).or_insert_with(|| {
+            LedgerInfoWithSignatures::new(vote_msg.ledger_info().clone(), HashMap::new())
+        });
+        vote_msg
+            .signature()
+            .clone()
+            .add_to_li(vote_msg.author(), li_with_sig);
```

Comment:
Also, has nothing to do with this PR, but I feel like it is better to call li_with_sig.add_signature() rather than SignatureWrapper.add_to_li() - SignatureWrapper doesn't need to know about LI.

Reply:
Also, nothing to do with this PR, but we should validate the signature against ValidatorVerifier (easier after #1149) before blindly added it to LI.

Reply:
signature verification should be done in network layer already?

Reply:
I agree. We should use `LedgerInfoWithSignatures::add_signature()`, it makes more sense. It doesn't work today with the CryptoProxy types, added a TODO to add the proper conversion in the CryptoProxy types.

Reply:
I think of network signature verification as a filtering mechanism, but when we compose a QC/TC, we should be certain at the point of creation - I like both.  To be fair, this is debatable and we should discuss it.